### ISSUE TITLE
onFinishUploaderHandler called before executeAction returns

### ISSUE THREAD

**Issue by GoogleCodeExporter** [bdshadow/gwtupload#29]
```
What steps will reproduce the problem?
1.  std single file upload
2.  pass the file to be uploaded
3.  on the servlet side add a delay of more than 1 second.

What is the expected output? 
onFinishUploaderHandler should not be called until the action is completed.

What do you see instead?

When I debug the code.  the client side onFinishUploaderHandler is called
before the break point is called.

What version of the product are you using? 

Latest.

On what operating system?

On Mac OS X 10.6.x
Please provide any additional information below.


```

Original issue reported on code.google.com by `rhicks...@gmail.com` on 8 Mar 2010 at 11:11

**Comment by GoogleCodeExporter** [comment]
```
What do you mean with 'add a delay', set the servlet slowUploads parameter 
equal to 1000?
Does it happen only in debug mode or it happens also in production?

Manolo
```

Original comment by `manuel.carrasco.m` on 10 Mar 2010 at 7:59
- Changed state: **Started**

**Comment by GoogleCodeExporter** [comment]
```
I put another blocking network call in the servlet:  executeAction.

This happens in production and debug.
```

Original comment by `rhicks...@gmail.com` on 10 Mar 2010 at 11:51

**Comment by GoogleCodeExporter** [comment]
```
I'm also having this problem. This is easy to reproduce. 

In Uploader.java the 'serverResponse = event.getResults();' returns nulls 
instead of 
'test' when I put a 'return "test" in executeAction().

Tried it with an image of approx. 200 kb

ps. Where should I put the delay as a workaround?



```

Original comment by `bono...@gmail.com` on 12 Mar 2010 at 6:06

**Comment by GoogleCodeExporter** [comment]
```
Well,

This is an expected behaviour. When the server finishes to receive the file it
updates the progress listener in session setting percentage as 100%. When the 
client
asks via Ajax asynchronously, it sets the widget as finished firing the events.
Finally the server sends the form response and the client updates the server
response. Normally the client sees the form response before the Ajax finished
response because of the delay in the client on periodic Ajax updates.
The reason of this is because I found some times, some browsers never fire the
`onSubmitHandler` and also because in many applications the client doesn't want 
to
wait while the server performs operations with the uploaded file.
You could override this behaviour extending `UploadListener` and overriding  
`getPercent`, also you have to override `getCurrentUploader` in your servlet, 
so you
could use it in your servlet in order not to return a 100% until it has 
finished.

Maybe I include this feature in the next release, making the behaviour 
configurable.

Let me know what you think

Manolo 
```

Original comment by `manuel.carrasco.m` on 28 Mar 2010 at 7:48

**Comment by GoogleCodeExporter** [comment]
```
The issue has been fixed in release 0.5.8-M2, please verify it.

Thank you for reporting.
Manolo
```

Original comment by `manuel.carrasco.m` on 6 Apr 2010 at 8:05
- Changed state: **Fixed**
